Feature/search main

letusHyun opened the pull request:
🚨 **Your checklist for this pull request** 

- [x]  Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!

- [x]  Check the commit's or even all commits' message styles matches our requested structure.

## 📌 [요약]
### 전체적으로 코드 리펙토링 진행
- CollectionViewCompositionalLayout을 관리하는 클래스를 새로 생성해서 VC의 로직을 덜어내었습니다.
- 오토레이아웃과 UI관련 매직넘버들을 enum Constants으로 만들었습니다.
- Search Modeling을 수정했습니다. 
- cellViewModel을 추가하고 input output패턴을 적용했습니다.

## ✨ [작업 내용]
##  Constants
매직넘버를 사용하는 객체에서 nested enum으로 Constants를 정의하고 사용함으로써 매직넘버를 제거했습니다.

## SearchModeling 수정
SearchVC는 collectionView의 다중 section을 적용하고 있습니다. 각 section은 서로 다른 하나의 Item 타입을 사용하고 있기때문에 Item을 enum으로 정의해주고 연관값으로 ItemModel값을 가진 CellViewModel을 전달하고 있습니다.

### 이전 코드
![image](https://github.com/IF-TG/iOS/assets/108918481/874a4026-8ec1-461c-9bd1-f79fadc51c29)
이전에는 section별 연관값으로 Item값과 section 값을 같이 전달했는데, 이는 불필요한 데이터까지 함께 전달하는 문제가 있었습니다. (SRP 위반)


### 현재 코드
![image](https://github.com/IF-TG/iOS/assets/108918481/74b94034-4269-47d1-8e18-a1cead1233b0)
리펙토링한 코드는 데이터 전달 시 Item값과 section값을 분리시켜서, 목적에 맞게 데이터를 전달할 수 있습니다. (SRP 준수)


### [Left:이전 모델링 결과, Right: 리펙토링 결과]
![image](https://github.com/IF-TG/iOS/assets/108918481/a9b63162-ea25-467e-99bb-69af2094ebbc)
![image](https://github.com/IF-TG/iOS/assets/108918481/61bcdba1-c3b7-44c6-8275-fa9e4cb07321)


##  cellViewModel 추가
각 cell마다 viewModel을 정의해서 input output패턴을 적용했습니다. 따라서 cell의 input은 cellViewModel로 흘러들어갑니다. 로직 처리 후, cell은 output을 binding처리합니다. cell이 CellViewModel과 직접 소통하므로, 기존에 있었던 cell과 VC의 deleagte패턴을 제거했습니다.

letusHyun committed:
[Refactor] 불필요한 코드 제거

letusHyun committed:
[Feat] section별 cell backgroundColor 설정

letusHyun committed:
[Feat] heartButton 기능 추가

letusHyun committed:
[Feat] heartButton 기능 추가

letusHyun committed:
Merge branch 'develop' of https://github.com/IF-TG/iOS into feature/search-main

letusHyun committed:
Merge branch 'develop' of https://github.com/IF-TG/iOS into feature/search-main

letusHyun committed:
[Style] 주석 수정, 생명주기 테스트, 코드 line 변경

letusHyun committed:
[Feat] input에 따라 버튼 color 변화

letusHyun committed:
[Feat] Repository Pattern

letusHyun committed:
[Refactor] '+' image 왼쪽에서 오른쪽으로 layout 변경

letusHyun committed:
[Refactor] 불필요 코드 제거

letusHyun committed:
[Refactor] color 수정

letusHyun committed:
[Refactor] SearchModel 수정

letusHyun committed:
[Refactor] configure인자로 CellVM 주입

letusHyun committed:
[Refactor] HeaderView Model 주입

letusHyun committed:
[Refactor] CompositionalLayout 클래스를 만들어서 구현로직을 VC와 분리

letusHyun committed:
[Refactor] dateFormat 수정

letusHyun committed:
[Refactor] enum의 연관값으로 CellVM배열 전달

letusHyun committed:
[Refactor] modeling 전체 수정

letusHyun committed:
[Style] 코드 포맷 변경

letusHyun committed:
[Style] 코드 포맷 변경

letusHyun committed:
[Feat] Cell과 CellViewModel간의 input ouput 패턴 추가

letusHyun committed:
[Feat] Cell이 CellViewModel을 참조

letusHyun committed:
[Refactor] 중복 코드 함수화

letusHyun committed:
[Feat] SearchCompositionalLayout 인터페이스 추가

letusHyun committed:
[Style] 불필요한 주석 제거

letusHyun committed:
[Refactor] 오토레이아웃 및 매직넘버를 Constants로 수정

letusHyun committed:
[Feat] Mock model 추가

letusHyun committed:
[Feat] Input을 프로토콜로 정의

letusHyun committed:
[Refactor] 오토레이아웃 매직 넘버를 Constants로 변경

letusHyun committed:
[Refactor] CoreText 사용해서 viewModel과 UIKit의 의존성 제거

letusHyun committed:
[Refactor] Constants로 수정

letusHyun committed:
[Comment] 일부 주석 수정

letusHyun committed:
[Feat] ViewBindCase 적용

letusHyun committed:
[Feat] ViewBindCase 적용

letusHyun committed:
[Refactor] 매직넘버를 Constants 로 변경

letusHyun committed:
[Refactor] SearchSectionType enum을 nested에서 전역으로 수정

letusHyun committed:
[Refactor] iOS 델리게이트 패턴 일관성을 위해 인자로 self 추가

letusHyun committed:
[Style] 불필요한 공백 제거

letusHyun committed:
[Feat] 델리게이트 패턴 기능, sectionType 프로퍼티 추가

letusHyun committed:
[Refactor] ViewModelCase 프로토콜 준수

letusHyun committed:
[Comment] input output 패턴 todo

letusHyun committed:
[Refactor] ViewModelInput의 중복이 발생해서 InputProtocol 적용

letusHyun committed:
[Refactor] heart button 커스텀 클래스 적용

각 cell마다 heartButton을 공통으로 사용하고 있어서 재사용성을 위해 커스텀 클래스로 추가했습니다.

letusHyun committed:
[Feat] FamousSpotCell의 ViewModel 추가

cell의 ViewModel을 만들어서 input output 패턴을 적용했습니다.

letusHyun committed:
[Refactor] 가시성을 위해 로직을 함수로 추출

section에 따라 header type을 지정해주는 로직을 private 함수로 추출했습니다.

letusHyun committed:
[Refactor] VC의 ViewModel에서 output로직 삭제

이제부터 CellViewModel에서 input에 대한 output을 관리하기 때문에, VC의 ViewModel에서 output 로직을 제거했습니다.

letusHyun committed:
[Remove] HeartButtonDelegate.swift 삭제

heartButton input을 VC가 아닌 cell에서 처리하므로, delegate패턴을 제거했습니다.

SHcommit commented:
OCP,,,

크으.. 붓캠끝나고 저도 SOLID 더 깊게 공부해보려고하는데 ㅎㅎ..
enum의 case와 그에 대한 값을 딕셔너리로 관리하는 거 어떤가요?..? +_+

https://marcduerst.com/2019/10/30/mapping-with-dictionary-instead-switch/

letusHyun committed:
[Style] 네이밍 mock으로 변경

letusHyun committed:
[Style] 함수 파라미터 네이밍 변경

letusHyun committed:
[Refactor] SearchModel의 modeling 수정

이전에는 section별 연관값으로 Item값과 section 값을 같이 전달했습니다. 하지만 이는 불필요한 데이터까지 함께 전달하는 문제가 있었습니다. (SRP 위반)
리펙토링한 코드는 데이터 전달 시 Item값과 section값을 분리시켜서, 목적에 맞게 데이터를 전달할 수 있습니다. (SRP 준수)

letusHyun committed:
[Style] 불필요한 주석 제거

letusHyun committed:
[Style] Search datasource 네이밍 변경

SHcommit commented:
형 미안,, 이번주가 붓캠마지막이라 밀린.. 요구사항 정주행중이야 ㅠㅠ..진짜 끝나고 나도 바로 개발 달려볼게

SearchSectionModel로 section의 enum을 프로퍼티로 추가하니까 vm에서 꺼내올 때도 깔끔하네요!!

추후에 시간이 난다면 viewModel에서 반환하는 getCellViewModels(in:), fetchHeaderTitle(in:), numberOfItemsInSections(in:) 이렇게 view에게 반환하는 함수들 또한 
뷰 컨트롤러에서 구체타입의 vm.geCellViewModels(in: ) 이렇게 호출하는게 아니라 protocol로 정의해서 viewController측은 아예 viewModel이 어느 viewModel 구체타입인지 모르게 리빌딩 하는것도 좋을 것 같습니다: ]

SHcommit commented on travelPlan/Source/Feature/Feed/View/PostSearch/PostSearchTagCell.swift:
```diff
@@ -42,22 +43,29 @@ final class PostSearchTagCell: UICollectionViewCell {
   }
   
   private func makeDeleteButton() -> UIButton {
     deleteButton = UIButton()
     guard let deleteButton = deleteButton else { return UIButton() }
-    deleteButton.setImage(UIImage(systemName: "xmark"), for: .normal)
+    deleteButton.setImage(UIImage(systemName: Constants.DeleteButton.imageName), for: .normal)
     deleteButton.tintColor = .yg.gray5
     deleteButton.addTarget(self, action: #selector(didTapDeleteButton), for: .touchUpInside)
     return deleteButton
   }
```
이 함수는 아직 호출을 안하나요?,,?

letusHyun replied:
makeDeleteButton함수 말씀하시는거라면 addSubViews의 분기 처리에서 호출합니다.

SHcommit commented on travelPlan/Source/Feature/Feed/View/PostSearch/PostSearchTagCell.swift:
```diff
@@ -72,14 +80,26 @@ extension PostSearchTagCell {
 // MARK: - Actions
 extension PostSearchTagCell {
   @objc private func didTapDeleteButton(_ button: UIButton) {
     delegate?.didTapDeleteButton(
       item: button.tag,
-      in: sectionType?.index ?? SectionType.recent.index
+      in: sectionType?.rawValue ?? SectionType.recent.rawValue
     )
   }
```
버튼 별로 태그를 다는 방법보다 좋은 방법이 없을까요?

letusHyun replied:
PostSearch는 Constants만 수정하고 아직 전반적인 리펙토링은 진행하지 못했었는데, 이 부분도 빠른 시일 내로 리펙토링하겠습니다.

SHcommit commented on travelPlan/Source/Feature/Feed/View/PostSearch/PostSearchTagCell.swift:
```diff
@@ -72,14 +80,26 @@ extension PostSearchTagCell {
     }
   }
+  
+  func setBackgroundColor(with section: Int) {
+    switch section {
+    case SectionType.recommendation.rawValue:
+      contentView.backgroundColor = UIColor(hex: Constants.Recommendation.backgroundColor)
+        .withAlphaComponent(Constants.Recommendation.alphaComponent)
+    case SectionType.recent.rawValue:
+      contentView.backgroundColor = .yg.veryLightGray
+    default:
+      contentView.backgroundColor = .white
+    }
+  }
```
외부시점에서 몇번째 section의 item index인지에 따라 dequeueReusbalecell메서드를 호출할텐데, 그때 같이 어느 sectionType 인지 알 수 없을까요? Cell내부에서는 외부에서 받은 값에 따라 바로바로 backgroundColor만 변경할 수 있도록 하는 방법은 별로일까요?

letusHyun replied:
ViewController 입장에서 Cell의 관심사를 알고있게 되면 SRP에 위배될 수 있다고 생각을 해서 Cell에서 sectionType을 처리해주었습니다. ViewController가 Cell의 관심사에 관여했을 때 장점이 있을까요?

SHcommit commented on travelPlan/Source/Feature/Feed/ViewController/PostSearchViewController.swift:
```diff
@@ -96,6 +107,10 @@ final class PostSearchViewController: UIViewController {
   }
+  
+  deinit {
+    print("DEBUG: \(String(describing: self)) deinit")
```
이번에 알게 된 건데,, os_log 나 Logger써볼까요?..

letusHyun replied:
좋아요 :) 좀 더 공부해보고 Logger로 수정하겠습니다!

SHcommit commented on travelPlan/Source/Feature/Feed/ViewModel/PostSearchViewModel.swift:
```diff
@@ -6,7 +6,8 @@
 import Combine
-import UIKit
+import Foundation
+import CoreText
```
호오,,

SHcommit commented on travelPlan/Source/Feature/Feed/ViewModel/PostSearchViewModel.swift:
```diff
@@ -144,29 +145,45 @@ extension PostSearchViewModel {
 // MARK: - Public Helpers
 extension PostSearchViewModel {
+  func sizeForItem(at indexPath: IndexPath) -> String {
+    return model[indexPath.section].items[indexPath.item]
```
한가지 TMI

indexPath는 컬랙션뷰일 떄는 item, 테이블 뷰일때는 row로 호출한다고합니다.

SHcommit commented on travelPlan/Source/Feature/Feed/ViewModel/PostSearchViewModel.swift:
```diff
@@ -144,29 +145,45 @@ extension PostSearchViewModel {
+    // 문자열 크기 계산
+    let textSize = attributedString
+      .boundingRect(
+        with: CGSize(width: CGFloat.greatestFiniteMagnitude, height: CGFloat.greatestFiniteMagnitude),
```
오 여기서 width에 leading, trailing spacing 뺴서 실제 text width 길이 측정 할 수도 있겠네요 +_+

SHcommit commented on travelPlan/Source/Feature/Search/Model/SearchFamousSpotModel.swift:
```diff
@@ -0,0 +1,37 @@
+
+import Foundation
+
+struct SearchFamousSpotModel {
+  typealias Identifier = Int
+  
+  let id: Identifier
+  let imageName: String?
+  let place: String
+  let location: String
+  let category: String
+  let isSelectedButton: Bool
+}
+
+// MARK: - Mock
+extension SearchFamousSpotModel {
+  static var mockModels: [SearchFamousSpotModel] = [
+    .init(id: 1, imageName: "tempProfile4",
+          place: "관광 장소명", location: "서울", category: "관광 카테고리", isSelectedButton: false),
+    .init(id: 1, imageName: "tempProfile4",
+          place: "관광 장소명", location: "서울", category: "관광 카테고리", isSelectedButton: false),
+    .init(id: 1, imageName: "tempProfile4",
+          place: "관광 장소명", location: "서울", category: "관광 카테고리", isSelectedButton: false),
+    .init(id: 1, imageName: "tempProfile4",
+          place: "관광 장소명", location: "서울", category: "관광 카테고리", isSelectedButton: false),
+    .init(id: 1, imageName: "tempProfile4",
+          place: "관광 장소명", location: "서울", category: "관광 카테고리", isSelectedButton: false),
+    .init(id: 1, imageName: "tempProfile4",
+          place: "관광 장소명", location: "서울", category: "관광 카테고리", isSelectedButton: false)
+  ]
+}
```
호오 +_+

SHcommit commented on travelPlan/Source/Feature/Search/Protocol/SearchCompositionalLayout.swift:
```diff
@@ -0,0 +1,12 @@
+
+import UIKit.UICollectionViewCompositionalLayout
+
+protocol SearchCompositionalLayout {
+  func createLayout() -> UICollectionViewCompositionalLayout
+}
```
나중에 컴포지셔널 한 수 배우겠습니다..

diffable data source도 정말 좋다고 합니다. 스냅샷으로 어느 cell이 바뀌었는지 해시값?으로 판별해서 갱신한다고하는데 애니메이션까지 스무스하게 된다고합니다 +_+

SHcommit commented on travelPlan/Source/Feature/Search/View/SearchBestFestivalCell.swift:
```diff
@@ -69,6 +65,63 @@ final class SearchBestFestivalCell: UICollectionViewCell {
   override func prepareForReuse() {
     super.prepareForReuse()
     thumbnailImageView.image = nil
+    heartButton.isSelected = false
+    periodLabel.text = nil
+    festivalLabel.text = nil
+    subscriptions.removeAll()
+  }
```
크으 좋습니다!!

SHcommit commented on travelPlan/Source/Feature/Search/View/SearchBestFestivalCell.swift:
```diff
@@ -69,6 +65,63 @@ final class SearchBestFestivalCell: UICollectionViewCell {
+  }
+  
+  internal func handleError(_ error: ErrorType) {
+    switch error {
+    case .fatalError:
+      print("DEBUG: Error fatalError in SeachBestFestivalCell")
+    case .unexpected:
+      print("DEBUG: Error unexpected in SeachBestFestivalCell")
+    case .networkError:
+      print("DEBUG: Error networkError in SeachBestFestivalCell")
+    }
```
나중에 서버에서 값을 제데로 못받았을 경우 retry관련 로직도 같이 짜보면 좋을 것 같습니다!!

SHcommit commented on travelPlan/Source/Feature/Search/View/SearchCollectionViewCompositionalLayout.swift:
```diff
@@ -0,0 +1,107 @@
+    )
+  }
+  
+  private func makeLayoutItem(
+    fractionalWidth: CGFloat,
+    fractionalHeight: CGFloat
+  ) -> NSCollectionLayoutItem {
+    return NSCollectionLayoutItem(layoutSize: .init(
+      widthDimension: .fractionalWidth(fractionalWidth),
+      heightDimension: .fractionalHeight(fractionalHeight)
+    ))
+  }
```
이야.. 고수,,

SHcommit commented on travelPlan/Source/Feature/Search/View/SearchFamousSpotCell.swift:
```diff
@@ -7,51 +7,61 @@
   }
   
+  private lazy var input = Input()
+  private var subscriptions = Set<AnyCancellable>()
```
요기도 prepareForReuse()로 dequeueReusableCell 시점 바인딩 중첩 조심해야합니다!!

letusHyun replied:
실수로 여기서는 초기화를 안해줬네요.. 수정했습니다. 감사합니다!

SHcommit commented on travelPlan/Source/Feature/Search/ViewModel/SearchFamousSpotCellViewModel.swift:
```diff
@@ -0,0 +1,92 @@
+  /// 성공 시 UI 변환, 실패 시, 변화 없음
+  private func saveButtonState(id: Int) -> Future<State, ErrorType> {
+    // networkTODO: - id값을 통해 서버에 데이터 저장을 요청하고, 성공 시 하트버튼의 색깔을 변경해야 합니다.
+    return Future { promise in
+      // fake network. 추후 네트워크 통신 이후, promise로 값을 방출해야 합니다.
+      DispatchQueue.global().asyncAfter(wallDeadline: .now() + 0.5) { [weak self] in
+        print("DEBUG: FakeNetwork 통신 성공!")
+        
+        self?.isSelectedButton.toggle()
+        promise(.success(.changeButtonColor))
+      }
+    }
```
키야.. !! 이거 testCode에서 작성해도 좋을 것 같습니다!!

letusHyun replied:
감사합니다!

SHcommit commented on travelPlan/Source/Feature/Search/ViewModel/SearchFamousSpotCellViewModel.swift:
```diff
@@ -0,0 +1,92 @@
+  func transform(_ input: Input) -> Output {
+    // 서버 저장 요청 후 응답에 따라 output이 달라집니다.
+    return input.didTapHeartButton
+      .flatMap { [weak self] _ in
+        guard let self = self else {
```
여기선 self 캡쳐가 일어나나요?

letusHyun replied:
flatMap 클로저 내에서는 guard let self 구문을 사용했기 때문에 self 캡쳐가 일어나는 것은 아니고, self.saveButtonState함수를 사용했기 떄문에 self캡쳐가 일어납니다. 하지만 [weak self] 키워드를 사용했기 때문에 강한 참조로 self 캡쳐를 하는 것이 아니라, 약한 참조로 self 캡쳐를 합니다.

 guard self = self구문을 사용하면 self를 강한 참조하는 코드가 되기 때문에 지연 메모리 해제가 발생할 수 있습니다. 여기서 지연 메모리 해제가 발생하는 것이 문제가 될 수 있다고 생각될 수도 있겠지만, 오히려 지연 메모리 해제를 해주어야 올바른 로직이라고 생각했습니다. 

한가지 예시를 들어보겠습니다. 성격이 급한 유저가 하트버튼을 누르고 곧바로 다른 화면으로 전환합니다. 하트버튼을 누르면 서버와 통신을 하는데, 이때 통신이 완료되고 하트버튼의 상태가 변경 되어야만 다음 화면으로 넘어갈 수 있도록 하는 것이 ux 관점에서 바라봤을 때 더 적합하다고 생각했기 때문에, guard let self = self 구문을 사용해서 지연 메모리 할당 해제를 사용했습니다. 

다른 의견이 있으시다면 알려주시면 감사하겠습니다!

SHcommit replied:
캡쳐는 클로저 내에서 클로저 외부의 값이나 참조객체를 사용할 떄 일어나는 것으로 알고 있는데, 이 경우 upstream의 input은 Void로 알고 있습니다. didTapHeartButton이라는 upstream publihser는 시간의 흐름에 따라 이벤트를 방출시키는 역할을 하고 있는 것으로 알고 있습니다. flatMap내부에 외부 컨텍스트를 사용하지 않아서 retain cycle이 발생되지 않을 것 같다는 생각이 듭니다.

함수를 호출하는데,, 함수가 강한 참조를 발생시키는지 확인이 필요할 것 같습니다

SHcommit commented on travelPlan/Source/Feature/Search/ViewModel/SearchViewModel.swift:
```diff
@@ -63,30 +90,50 @@ extension SearchViewModel: ViewModelCase {
 extension SearchViewModel {
-  func fetchModel(in section: Int) -> SearchSectionItemModel.SearchSection {
-    return models[section]
+  func getCellViewModels(in section: Int) -> SearchItemType {
+    return dataSource[section].itemType
```
아하 이렇게 모델을 넘겨서 collectionView(_:cellForRowAt:)에서 switch로 비교하시는군요

SHcommit reviewed (approved):
붓캠 다녀올 사이에.. 빡고수가 되셨네요.... +_+ 

확인했습니다.
